### PR TITLE
iox-eclipse-iceoryx#2040 support for iox::string in UnixDomainSocket

### PR DESCRIPTION
**Pull request by lucabart97**
By default, the UnixDomainSocket class relies on std::string objects to send and receive data, as defined by its interface. However, this approach can potentially lead to dynamic memory allocation when handling larger data payloads that exceed the stack space optimization (SSO) limit.

To address this issue, an alternative API has been introduced that enables data transmission and reception using iox::string. This approach allows for direct data manipulation within stack memory, effectively eliminating the need for dynamic memory allocation.

## Pre-Review Checklist for the PR Author

1. [x] Add a second reviewer for complex new features or larger refactorings
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Reference: https://github.com/eclipse-iceoryx/iceoryx/issues/2040
- Reference: https://github.com/eclipse-iceoryx/iceoryx/issues/1693

### CODE REVIEW

**Review by elBoberido — COMMENTED** [review]
Did not yet look at the tests.

Unfortunately there are some merge conflicts due to the ongoing restructuring of iceoryx hoofs. I hope they are not too severe.

Could you also refer to #1693 in the PR description.

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp [hunk 1]
```diff
@@ -344,21 +364,20 @@ expected<std::string, IpcChannelError> UnixDomainSocket::timedReceive(const unit
...
-
     if (recvCall.has_error())
     {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        msg[0] = 0;
         return err(errnoToEnum(recvCall.error().errnum));
     }
-    return ok<std::string>(&message[0]);
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    msg[static_cast<uint64_t>(recvCall->value) + NULL_TERMINATOR_SIZE] = 0;
```
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp]: To avoid duplicated code, I add a boolean option
```
expected<void, PosixIpcChannelError> sendImpl(const char* buf, uint64_t bufSize, const units::Duration& timeout, const bool hasTerminator) const noexcept;

expected<uint64_t, PosixIpcChannelError> receiveImpl(char* buf, uint64_t maxBufSize, const units::Duration& timeout, const bool addTerminator) const noexcept;
```

File: iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp [hunk 2]
```diff
@@ -344,21 +369,27 @@ UnixDomainSocket::timedReceive(const units::Duration& timeout) const noexcept
...
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+            buf[0] = 0;
+        }
         return err(errnoToEnum(recvCall.error().errnum));
     }
-    return ok<std::string>(&message[0]);
+    if (addTerminator)
+    {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        buf[static_cast<uint64_t>(recvCall->value) + NULL_TERMINATOR_SIZE] = 0;
```
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:389]: I agree about option 2, I hope that I do not need to change a lot of code!

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp]: Here is my proposal:

```c++
enum class Termination {
NONE,
NULL_TERMINATOR, // STRING or STRING_TERMINATOR might also work
};

template <Typename Type>
expected<void, PosixIpcChannelError> sendImpl(const Type* msg, uint64_t msgSize, const units::Duration& timeout, const Termination terminator) const noexcept;

template <Typename Type>
expected<uint64_t, PosixIpcChannelError> receiveImpl(Type* msg, uint64_t maxBufSize, const units::Duration& timeout, const Termination terminator) const noexcept
```

The Template pointer is needed in the future if we want to send a vector of ```std::byte```, ```unit8_t``` etc.

If ```sendImpl``` is called with ```NULL_TERMINATOR```, the termination size (1) is added to ```msgSize``` to send also the terminator.

if ```receiveImpl``` is called with ``NULL_TERMINATOR```, the last buffer data arrived is checked has a terminator, otherwise, I return an error.

[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp]: This would work. With the additional template parameter there is a new option. If we would assume that a `char` type would always be null-terminated, we could omit the enum and just check for the type. This makes it less flexible though since it will not be possible to send a vector of chars anymore.

I think your proposal should work better. We also need to keep in mind to not make the code too complex just to prevent a bit of code duplication but up until now it is still okay.

**Comment by elBoberido** [comment]
@lucabart97 sorry for letting you wait so long. I was quite busy this week and will have a look at the PR tomorrow.

**Comment by lucabart97** [comment]
@elBoberido no problem, take your time!

**Review by elBoberido — COMMENTED** [review]
I just skimmed through the tests but with the new behavior of having buffers which are potentially smaller than the message, new tests need to check this behavior.

I think a test with a `iox::string` full of e.g. `a` will unveil an off by one error in the `receive` method.

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            msg[0] = 0;
+        }
+        return err(errnoToEnum(recvCall.error().errnum));
+    }
+    if (terminator == Termination::NULL_TERMINATOR)
+    {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        if (msg[static_cast<uint64_t>(recvCall->value)] != 0)
+        {
+            return err(PosixIpcChannelError::INTERNAL_LOGIC_ERROR);
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:108]: > A few lines further above you add a null-termination in case of an error. Can you share your thoughts why you decided against it in this case?

Forget about it. I totally forgot that I proposed to not write a null-termination at the end but return an error.

The problem I overlooked is just that we might break an invariant of the `iox::string` since some data might be written to the buffer. The size would still be zero but the value at index 0 would not be the null-termination. Can you please write a 0 to `str[0]` right before `return 0;` in `timedReceive(iox::string<N>& ...`.

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+#ifndef IOX_HOOFS_POSIX_IPC_UNIX_DOMAIN_SOCKET_INL
+#define IOX_HOOFS_POSIX_IPC_UNIX_DOMAIN_SOCKET_INL
+
+#include "iox/posix_call.hpp"
+#include "iox/unix_domain_socket.hpp"
+
+namespace iox
+{
+template <typename Type>
+expected<void, PosixIpcChannelError> UnixDomainSocket::sendImpl(const Type* msg,
```
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:25]: I got some issues using ```not_null```.
This is a particular case where the functions are templates: the compiler does not deduce the type. Do you have any suggestions?

```
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp: In instantiation of ‘iox::UnixDomainSocket::timedReceive(const iox::units::Duration&) const::<lambda(auto:8*, auto:9)> [with auto:8 = char; auto:9 = iox::BufferInfo; uint64_t = long unsigned int]’:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/functional/include/iox/detail/function_ref.inl:45:90:   required from ‘iox::function_ref<ReturnType(ArgTypes ...)>::function_ref(CallableType&&) [with CallableType = iox::UnixDomainSocket::timedReceive(const iox::units::Duration&) const::<lambda(auto:8*, auto:9)>; <template-parameter-2-2> = void; ReturnType = long unsigned int; ArgTypes = {char*, iox::BufferInfo}]’
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:308:6:   required from here
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:302:16: error: no matching function for call to ‘iox::UnixDomainSocket::timedReceiveImpl(char*&, const uint64_t&, const iox::units::Duration&, iox::UnixDomainSocket::Termination) const’
  302 |         result = this->timedReceiveImpl(str, info.total_size, timeout, Termination::NULL_TERMINATOR);
In file included from /home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:19:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/unix_domain_socket.hpp:163:46: note: candidate: ‘template<class Type> iox::expected<long unsigned int, iox::PosixIpcChannelError> iox::UnixDomainSocket::timedReceiveImpl(iox::not_null<Type*>, uint64_t, const iox::units::Duration&, iox::UnixDomainSocket::Termination) const’
  163 |     expected<uint64_t, PosixIpcChannelError> timedReceiveImpl(not_null<Type*> msg,
      |                                              ^~~~~~~~~~~~~~~~
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/unix_domain_socket.hpp:163:46: note:   template argument deduction/substitution failed:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:302:16: note:   mismatched types ‘iox::not_null<Type*>’ and ‘char*’
  302 |         result = this->timedReceiveImpl(str, info.total_size, timeout, Termination::NULL_TERMINATOR);
```

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                              .failureReturnValue(ERROR_CODE)
+                              .ignoreErrnos(EWOULDBLOCK)
+                              .evaluate();
+
+    if (setsockoptCall.has_error())
+    {
+        return err(errnoToEnum(setsockoptCall.error().errnum));
+    }
+    uint64_t msgSizeToSend = msgSize;
+    if (terminator == Termination::NULL_TERMINATOR)
```
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:52]: The ```terminator``` value is not a constexpr. 
I also tried with the compiler to be sure, and this is the error I got:
```
In file included from /home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/unix_domain_socket.hpp:221,
                 from /home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:19:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl: In member function ‘iox::expected<void, iox::PosixIpcChannelError> iox::UnixDomainSocket::timedSendImpl(const Type*, uint64_t, const iox::units::Duration&, iox::UnixDomainSocket::Termination) const’:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:53:61: error: ‘terminator’ is not a constant expression
   53 |     if constexpr (terminator == Termination::NULL_TERMINATOR)
      |
```

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        return err(errnoToEnum(setsockoptCall.error().errnum));
+    }
+
+    auto recvCall = IOX_POSIX_CALL(iox_recvfrom)(m_sockfd, msg, maxMsgSize, 0, nullptr, nullptr)
+                        .failureReturnValue(ERROR_CODE)
+                        .suppressErrorMessagesForErrnos(EAGAIN, EWOULDBLOCK)
+                        .evaluate();
+    if (recvCall.has_error())
+    {
+        if (terminator == Termination::NULL_TERMINATOR)
```
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:96]: same

File: iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp [hunk 7]
```diff
@@ -276,89 +277,40 @@ expected<void, PosixIpcChannelError> UnixDomainSocket::send(const std::string& m
...
-    auto sendCall = IOX_POSIX_CALL(iox_sendto)(m_sockfd, msg.c_str(), msg.size() + NULL_TERMINATOR_SIZE, 0, nullptr, 0)
-                        .failureReturnValue(ERROR_CODE)
-                        .evaluate();
-
-    if (sendCall.has_error())
-    {
-        return err(errnoToEnum(sendCall.error().errnum));
-    }
-    return ok();
+    return sendImpl(msg.c_str(), msg.size(), timeout, Termination::NULL_TERMINATOR);
```
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:286]: sure!

**Comment by lucabart97** [comment]
> I just skimmed through the tests but with the new behavior of having buffers which are potentially smaller than the message, > new tests need to check this behavior.

> I think a test with a iox::string full of e.g. a will unveil an off by one error in the receive method.

I think the tests starting from this line do what do you suggest to do:

https://github.com/lucabart97/iceoryx/blob/1fb9063e91dc4602e2d3bcc4febc12dc0d37074e/iceoryx_hoofs/test/moduletests/test_posix_ipc_unix_domain_sockets.cpp#L401

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 5 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:52]: Oh, right. `Termination` must become a template parameter.

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 6 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:96]: I think it should work when it becomes a template parameter

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 4 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl]: Oh, you have to use `not_null<Type>` instead of `not_null<Type*>`. At least with clang it worked when passing a `not_null(str)` to the method.

**Comment by elBoberido** [comment]
> > I just skimmed through the tests but with the new behavior of having buffers which are potentially smaller than the message, > new tests need to check this behavior.
> 
> > I think a test with a iox::string full of e.g. a will unveil an off by one error in the receive method.
> 
> I think the tests starting from this line do what do you suggest to do:
> 
> https://github.com/lucabart97/iceoryx/blob/1fb9063e91dc4602e2d3bcc4febc12dc0d37074e/iceoryx_hoofs/test/moduletests/test_posix_ipc_unix_domain_sockets.cpp#L401 (401~509)

Not quite. Currently the data array in `iox::string` is zero initialized. This means that an off by one error might slip through.

There should be a test with a `message_t msg;` which is set to e.g. `a` for the full length and there also needs to be a tests with an `iox::string` with a capacity smaller than the one from `message_t`.

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:52]: My suggestion is to keep it as is, the function signature is simple, otherwise, when you call the method you need to pass the type and the terminator as a template parameter, overcomplicating the function call.

I also see no better performance to move in a template parameter: the function is a template and is not compiled and linked so that the terminator will be hardcoded to ```false``` or ```true```, the compiler will do the same optimization.

What do you think about it?

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl]: If I remember correctly, I tried both signatures, but the compilation failed.
This is the first case in the entire code that ```not_null``` is used with a  template parameter. 
I'll check better, but I'm pretty sure about that.

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 5 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:52]: Oh, I was not clear enough. I meant a non-type template parameter. The usage would then be `timedSendImpl<Termination::NULL_TERMINATOR>(msg, size, timeout)`. But you are right, the compiler should be able to optimize this anyway. There might be some tools which don't get it right and complain but that is something for our future selves. I'll leave it to you.

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 4 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl]: Sorry, you are right. It can be `not_null<Type*>`. I guess `Type` being a template parameter somehow prevents the implicit conversion to `not_null`. By explicitly using the ctor the problem is gone :)

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl]: Specified the template parameter in the function call the code works

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl]: Specified the template parameter in the function call the code works

File: iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucabart97) [reply iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl]: Given that I need to specify the msg type to use the```not_null``` constructor, I also added it to the template parameter list

**Comment by lucabart97** [comment]
I added a test with the ```iox::string``` full of ```a``` and different tests with different ```iox::string``` size in send and receive

**Comment by elBoberido** [comment]
Will look at it tomorrow more closely. Just skimmed over the changes and found a small stylistic issue. For non type template parameter we usually use upper case for the name.

File: iceoryx_hoofs/test/moduletests/test_posix_ipc_unix_domain_sockets.cpp [hunk 8]
```diff
@@ -374,6 +414,258 @@ TEST_F(UnixDomainSocket_test, SuccessfulCommunicationOfMultipleMessagesWithSendA
...
+        {""}, [&](auto& msg) { return client.send(msg); }, [&](auto& msg) { return server.timedReceive(msg, 1_ms); });
+}
+
+TEST_F(UnixDomainSocket_test, SuccessfulCommunicationOfFullLenghtMessageWithSendAndReceiveMsg)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "dc059e53-3d31-4ad5-93be-bbf0a1c0425d");
+    auto message = memsetMessage<message_t>('a');
+    successfulSendAndReceiveMsg<message_t, message_t>(
+        {message}, [&](auto& msg) { return client.send(msg); }, [&](auto& msg) { return server.receive(msg); });
+}
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_posix_ipc_unix_domain_sockets.cpp:583]: I think I was a bit unclear with my `a test with a iox::string full of e.g. a` comment. It's great to have this tests with a full buffer since it is an edge case but can you add another one.

My initial intention was to have a receive buffer full of `a` and receive something small, e.g. `all glory to the hypnotoad`. The reason for this is that the whole buffer of the `iox::string` is currently set to 0. We should rethink this decision but it also means that an off by one error is easily overlooked due to the zeroed buffer. It doesn't mean that there is one currently there but the tests are also there to prevent one in a future refactoring.

What we need is basically this
```cpp
successfulSendAndReceiveMsg<message_t, message_t>(
    {"All glory to the hypnotoad"}, [&](auto& msg) { return client.send(msg); }, [&](auto& msg) {
        msg = memsetMessage<message_t>('a');
        return server.receive(msg);
    });
```
It is sufficient to have it only once and does not need to be repeated for timedSend/timedReceive.

**Review by elBoberido — COMMENTED** [review]
Looks good. Just a small nitpick and one additional tests and from my point of view this can be merged. :)

Btw, next Tuesday is the next developer meetup and the week after there is another one to introduce the Rust based iceoryx2. You are welcome to join.

**Comment by lucabart97** [comment]
@elBoberido 
Done!

Sorry for the delay and thank you for the invite, but last week I was on holiday!

File: iceoryx_hoofs/test/moduletests/test_posix_ipc_unix_domain_sockets.cpp (hunk 8 above)
Comment (lucabart97) [reply iceoryx_hoofs/test/moduletests/test_posix_ipc_unix_domain_sockets.cpp:583]: https://github.com/lucabart97/iceoryx/blob/641d81d534936565898e1f9a81e70b70a9754c56/iceoryx_hoofs/test/moduletests/test_posix_ipc_unix_domain_sockets.cpp#L615-L625

**Review by elBoberido — APPROVED** [review]

**Comment by elBoberido** [comment]
Thanks for your patience :)

There is usually one meetup on the first Tuesday each month. Your welcome to join whenever it suits you :)
